### ISSUE TITLE
Hanging on swallowed error

### ISSUE THREAD

**Issue by morgoth** [socketry/async#161]
This simple code hangs:
```ruby
Async do |task|
  task.async do
    # raise TypeError # this properly raises an error
    sleep nil
  end
end.wait
```

```ruby
sleep nil
(irb):62:in `sleep': can't convert NilClass into time interval (TypeError)
```
so when I raise TypeError by hand it properly interrupts and report an error, but hangs on the `sleep nil` which standalone raises TypeError

**Comment by ioquatix** [comment]
Thanks @bruno- this is an interesting compatibility issue.

Since `sleep(nil)` is invalid in Ruby, maybe we could consider introducing this behaviour to make it consistent. Handling zero argument case rather than default argument = nil is less efficient in Pure Ruby than in C.

**Comment by ioquatix** [comment]
I think `sleep()` and `sleep(nil)` should be considered the same, but in CRuby it currently isn't. I don't have a solution. If you write this:

```ruby
def sleep(time = nil)
  if time
    sleep_for_time(time)
  else
    sleep_forever
  end
end
```

It's impossible to know if the time was nil because it was the default or the user provided nil. It's hard to replicate the CRuby interface.

**Comment by ioquatix** [comment]
https://bugs.ruby-lang.org/issues/19094

**Comment by ioquatix** [comment]
It was accepted by Matz, so this will be fixed in Ruby 3.3: https://github.com/ruby/ruby/pull/7484
